### PR TITLE
adding api to trigger throttled compaction immediately

### PR DESCRIPTION
**Pull request by samadsid7**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,7 +33,7 @@
          <url>https://github.com/flipkart-incubator/hbase-compactor</url>
          <tag>HEAD</tag>
     </scm>
-    
+
```
Comment (diptam-in) [inline pom.xml:36]: unnecessary changes

File: src/main/java/com/flipkart/yak/config/CompactionContext.java [hunk 2]
```diff
@@ -56,10 +56,14 @@ public int hashCode() {
 
     @Override
     public void validate() throws ConfigurationException {
-        if (tableName==null && nameSpace == null && rsGroup == null) {
+        validateTable(tableName, nameSpace, rsGroup);
+    }
+
+    static void validateTable(String tableName, String nameSpace, String rsGroup) throws ConfigurationException {
+        if (tableName == null && nameSpace == null && rsGroup == null) {
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/config/CompactionContext.java:63]: whats the requirement for this ?

File: src/main/java/com/flipkart/yak/config/PromptSchedule.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.SuperBuilder;
+import lombok.extern.jackson.Jacksonized;
+
+@Data
+@SuperBuilder
+@RequiredArgsConstructor
+@Jacksonized
+public class PromptSchedule {
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/config/PromptSchedule.java:12]: its not really a schedule. 
PromptCompactionLifespan you can say

File: src/main/java/com/flipkart/yak/config/loader/AbstractConfigWriter.java [hunk 4]
```diff
@@ -11,5 +11,6 @@ public abstract class AbstractConfigWriter <Resource> {
     public abstract boolean storeProfile(Resource resource, CompactionProfileConfig compactionProfileConfig);
     public abstract boolean storeContext(Resource resource, CompactionContext compactionContext);
     public abstract boolean deleteContext(Resource resource, CompactionContext compactionContext);
+    public abstract boolean deleteStaleContexts(Resource resource) throws ConfigurationException;
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/config/loader/AbstractConfigWriter.java:14]: rename it to deleteAllStaleContexts

File: src/main/java/com/flipkart/yak/config/zkstore/ZKConfigStoreWriter.java [hunk 5]
```diff
@@ -94,4 +100,29 @@ public boolean deleteContext(CuratorFramework zooKeeper, CompactionContext compa
...
+                    }
+                } catch (Exception  e) {
+                    log.error("could not get data from zookeeper path {}: {}", path, e.getMessage());
+                    throw new RuntimeException(e);
+                }
+            });
+            return true;
+        } catch (Exception e) {
+            log.error("Could not load zookeeper data: {}", e.getMessage());
+            throw new ConfigurationException(e);
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/config/zkstore/ZKConfigStoreWriter.java:125]: RuntimeException

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java [hunk 6]
```diff
@@ -39,15 +41,38 @@ public void init(CompactionContext context) throws ConfigurationException {
 
     @Override
     public void run() {
-        Thread.currentThread().setName(compactionContext.getTableName()+"-"+compactionContext.getNameSpace());
+        if(compactionSchedule.isPrompt()) {
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java:44]: move this logic to a separate method

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java [hunk 7]
```diff
@@ -39,15 +41,38 @@ public void init(CompactionContext context) throws ConfigurationException {
...
+        }
         MDC.put("JOB", Thread.currentThread().getName());
         log.info("starting compact-cron for : {}", this.getCompactionContext());
         while(true) {
+
+            /*
+            If prompt job , and its life cycle has already ended , exit
+             */
+            if(compactionSchedule.isPrompt() && ScheduleUtils.hasExpired(compactionSchedule, Instant.now())) {
+                Thread.currentThread().interrupt();
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java:57]: Why do you need to interrupt it ? 
Can it die its own death naturally?

If it has crosses timestamp, it means lifespan is over. Does it need to interrupt anything ?

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java [hunk 8]
```diff
@@ -39,15 +41,38 @@ public void init(CompactionContext context) throws ConfigurationException {
...
+                break;
+            }
             MonitorService.resetMeterValue(compactionExecutable.getClass(), compactionContext, "success");
             MonitorService.resetMeterValue(compactionExecutable.getClass(), compactionContext, "failure");
             this.compactionManager.checkAndStart();
+
+            /*
+            If prompt job , exit from schedule after completing compaction
+             */
+            if(compactionSchedule.isPrompt()) {
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java:68]: Why do you need it again ??

File: src/main/java/com/flipkart/yak/util/PromptCompactionUtil.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.flipkart.yak.commons.ScheduleUtils;
+import com.flipkart.yak.config.CompactionContext;
+import com.flipkart.yak.config.CompactionSchedule;
+import com.flipkart.yak.config.PromptCompactionRequest;
+import com.flipkart.yak.config.PromptSchedule;
+
+public class PromptCompactionUtil {
+
+    public static CompactionContext getCompactionContext(PromptCompactionRequest promptCompactionRequest) {
```
Comment (diptam-in) [inline src/main/java/com/flipkart/yak/util/PromptCompactionUtil.java:11]: Can it be part of any other Utility class. 
Util class is just a toolBox, you don't need granular specialisation for Util classes.

**Review by diptam-in — CHANGES_REQUESTED** [review]
Please address comments

File: src/main/java/com/flipkart/yak/config/CompactionContext.java (hunk 2 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/config/CompactionContext.java:63]: Its being shared with PromptCompactionRequest Class as well , hence decided to share it between the two classes through functions

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java (hunk 6 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java:44]: done

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java (hunk 8 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java:64]: modified now

File: src/main/java/com/flipkart/yak/config/PromptSchedule.java (hunk 3 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/config/PromptSchedule.java]: done

File: src/main/java/com/flipkart/yak/config/loader/AbstractConfigWriter.java (hunk 4 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/config/loader/AbstractConfigWriter.java]: done

File: src/main/java/com/flipkart/yak/config/zkstore/ZKConfigStoreWriter.java (hunk 5 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/config/zkstore/ZKConfigStoreWriter.java]: done

File: src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java (hunk 7 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/core/ThreadedCompactionJob.java]: hmmm earlier took safety measure to kill it once job is done .
But it can die its own death naturally as well , have modified it now.

File: src/main/java/com/flipkart/yak/util/PromptCompactionUtil.java (hunk 9 above)
Comment (samadsid7) [reply src/main/java/com/flipkart/yak/util/PromptCompactionUtil.java]: moved to commons

File: pom.xml (hunk 1 above)
Comment (samadsid7) [reply pom.xml]: corrected

**Review by diptam-in — APPROVED** [review]
